lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -0,0 +1,1009 @@
+        // Don't record instance breadcrumbs.  We're only looking for the final
+        // target of an NLA.
+        if (anno.getClassAttr() == nonLocalString)
+          continue;
+        targetMap[nlaRef.getAttr()] = OpAnnoTarget(op);
+      }
+    }
+
+    auto portAnnotations = op->getAttrOfType<ArrayAttr>("portAnnotations");
+    if (!portAnnotations)
+      return;
+    auto module = dyn_cast<FModuleOp>(op);
+    auto mem = dyn_cast<MemOp>(op);
+    for (auto pair : llvm::enumerate(portAnnotations))
+      for (auto anno : AnnotationSet(pair.value().cast<ArrayAttr>()))
+        if (auto nlaRef = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal")) {
+          // Breadcrumbs don't appear on port annotations, so we can skip the
+          // class check that we have above.
+          if (module)
+            targetMap[nlaRef.getAttr()] = PortAnnoTarget(module, pair.index());
+          else if (mem)
+            targetMap[nlaRef.getAttr()] = PortAnnoTarget(mem, pair.index());
+          else
+            llvm_unreachable("unknown operaiton with ports");
+        }
+  }
+
+  /// This fixes up connects when the field names of a bundle type changes.  It
+  /// finds all fields which were previously bulk connected and legalizes it
+  /// into a connect for each field.
+  void fixupConnect(ImplicitLocOpBuilder &builder, Value dst, Type dstType,
+                    Value src, Type srcType) {
+    // If its not a bundle type, the types are guaranteed to be unchanged.  If
+    // it is a bundle type, we would rather bulk-connect the values instead of
+    // decomposing the connect if the type is unchanged.
+    if (dstType == dst.getType() && srcType == src.getType()) {
+      builder.create<ConnectOp>(dst, src);
+      return;
+    }
+    // It must be a bundle type and the field name has changed. We have to
+    // manually decompose the bulk connect into a connect for each field.
+    auto dstBundle = dstType.cast<BundleType>();
+    auto srcBundle = srcType.cast<BundleType>();
+    for (unsigned i = 0; i < dstBundle.getNumElements(); ++i) {
+      auto dstField = builder.create<SubfieldOp>(dst, i);
+      auto srcField = builder.create<SubfieldOp>(src, i);
+      if (!dstBundle.getElement(i).isFlip)
+        fixupConnect(builder, dstField, dstBundle.getElementType(i), srcField,
+                     srcBundle.getElementType(i));
+      else
+        fixupConnect(builder, srcField, srcBundle.getElementType(i), dstField,
+                     dstBundle.getElementType(i));
+    }
+  }
+
+  /// This fixes up a partial connect when the field names of a bundle type
+  /// changes.  It finds all the fields which were previously connected and
+  /// replaces them with new partial connects.
+  using LazyValue = llvm::function_ref<Value(ImplicitLocOpBuilder &)>;
+  // NOLINTNEXTLINE(misc-no-recursion)
+  void fixupPartialConnect(ImplicitLocOpBuilder &builder, LazyValue dst,
```

Comment:
Shouldn't the bundle types be compatible for the hash to match, in that each field type matches?   The you know field 0 -> field 0, field 1 -> field 1, etc.  You shouldn't need to search by name.  It seems, you should be able to do a bitcast and be done with it.